core/src/main/java/org/mule/registry/AbstractRegistryBroker.java

```diff
@@ -169,68 +227,111 @@ public boolean isRemote()
+
+    @Override
     public <T> Collection<T> lookupObjectsForLifecycle(Class<T> type)
     {
-        Collection<T> objects = new ArrayList<T>();
+        Collection<Registry> registries = getRegistries();
+        if (registries.isEmpty())
+        {
+            return limbo.lookupObjectsForLifecycle(type);
+        }
 
-        Iterator it = getRegistries().iterator();
-        while (it.hasNext())
+        Collection<T> objects = new ArrayList<>();
+        for (Registry registry : registries)
         {
-            objects.addAll(((Registry) it.next()).lookupObjectsForLifecycle(type));
+            objects.addAll(registry.lookupObjectsForLifecycle(type));
         }
+
         return objects;
     }
 
-       
+    /**
+     * {@inheritDoc}
+     */
+    @Override
     public void registerObject(String key, Object value) throws RegistrationException
     {
-        registerObject(key, value, null);
-    }
-
-    public void registerObject(String key, Object value, Object metadata) throws RegistrationException
-    {
-        Iterator it = getRegistries().iterator();
-        Registry reg;
-        while (it.hasNext())
+        Collection<Registry> registries = getRegistries();
+        if (registries.isEmpty())
+        {
+            limbo.registerObject(key, value);
+        }
+        else
         {
-            reg = (Registry) it.next();
-            if (!reg.isReadOnly())
+            for (Registry registry : registries)
             {
-                reg.registerObject(key, value, metadata);
-                break;
+                if (!registry.isReadOnly())
+                {
+                    registry.registerObject(key, value);
+                    break;
+                }
             }
         }
     }
 
-    public void registerObjects(Map objects) throws RegistrationException
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @Deprecated
+    public void registerObject(String key, Object value, Object metadata) throws RegistrationException
     {
-        Iterator it = objects.keySet().iterator();
-        Object key;
-        while (it.hasNext())
-        {
-            key = it.next();
-            registerObject((String) key, objects.get(key));
-        }
+        registerObject(key, value);
     }
 
-    public void unregisterObject(String key) throws RegistrationException
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void registerObjects(Map objects) throws RegistrationException
     {
-        unregisterObject(key, null);
+        for (Entry<String, Object> entry : (Set<Entry<String, Object>>) objects.entrySet())
+        {
+            registerObject(entry.getKey(), entry.getValue());
+        }
     }
 
-    public void unregisterObject(String key, Object metadata) throws RegistrationException
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Object unregisterObject(String key) throws RegistrationException
     {
-        Iterator it = getRegistries().iterator();
```

Comment:
The ObjectLimbo concept makes this class more ugly than before, because every method is doing different things depending on the number of available registries